File: Multiple text file word replacer script.py
import os

def reemplazar_texto_en_archivo(ruta_archivo_val, expresiones_val, reemplazos_val):
    # Reemplaza las expresiones exactas y sus versiones en mayúsculas en un archivo de texto.
    try:
        with open(ruta_archivo_val, 'r', encoding='utf-8') as archivo_val:
            contenido_val = archivo_val.read()

        contenido_modificado_val = contenido_val
        modificado_flag = False  # Bandera para verificar si se realizó alguna modificación

        # Reemplazar cada expresión y su versión en mayúsculas
        for expr_val, repl_val in zip(expresiones_val, reemplazos_val):
            if expr_val in contenido_val or expr_val.upper() in contenido_val:
                contenido_modificado_val = contenido_modificado_val.replace(expr_val, repl_val)
                contenido_modificado_val = contenido_modificado_val.replace(expr_val.upper(), repl_val.upper())
                modificado_flag = True  # Se realizó al menos una modificación

        return contenido_modificado_val if modificado_flag else None  # Devolver el contenido modificado solo si hubo cambios

    except (UnicodeDecodeError, PermissionError, IOError) as e:
        print(f"Error processing file {ruta_archivo_val}: {e}")
        return None

def buscar_y_reemplazar_en_directorio(directorio_base, expresiones_val, reemplazos_val, directorio_salida):
    # Recorre recursivamente un directorio y reemplaza las expresiones en archivos de texto.
    contador_archivos = 0  # Contador de archivos modificados
    for directorio_actual, _, archivos_val in os.walk(directorio_base):
        for archivo_val in archivos_val:
            ruta_completa = os.path.join(directorio_actual, archivo_val)
            # Verificar si es un archivo de texto (puedes agregar más extensiones si es necesario)
            if archivo_val.endswith('.txt'):
                print(f"Processing file: {ruta_completa}")
                # Obtener el contenido modificado
                contenido_modificado_val = reemplazar_texto_en_archivo(ruta_completa, expresiones_val, reemplazos_val)
                if contenido_modificado_val is not None:
                    # Crear la ruta de salida manteniendo la estructura de directorios
                    ruta_relativa = os.path.relpath(directorio_actual, directorio_base)
                    subdirectorio_salida = os.path.join(directorio_salida, ruta_relativa)
                    os.makedirs(subdirectorio_salida, exist_ok=True)  # Crear subdirectorios si no existen
                    ruta_archivo_salida = os.path.join(subdirectorio_salida, archivo_val)
                    # Guardar el archivo modificado en la carpeta de salida
                    with open(ruta_archivo_salida, 'w', encoding='utf-8') as archivo_salida:
                        archivo_salida.write(contenido_modificado_val)
                    contador_archivos += 1  # Incrementar el contador de archivos modificados

    return contador_archivos  # Devolver el número de archivos modificados

def obtener_directorio_salida_unico(base_dir):
    # Generar un nombre de directorio único para evitar sobreescribir
    contador = 1
    directorio_salida = base_dir
    while os.path.exists(directorio_salida):
        directorio_salida = f"{base_dir} ({contador})"
        contador += 1
    return directorio_salida

def main():
    while True:  # Bucle infinito
        # Solicitar el directorio de entrada
        while True:
            directorio_base = input("Enter directory: ")
            if os.path.exists(directorio_base):
                break
            print("Wrong directory")

        # Solicitar el número de expresiones a modificar
        while True:
            try:
                num_expresiones = int(input("Enter number of expressions: "))
                break
            except ValueError:
                print("Wrong format")

        expresiones_val = []
        reemplazos_val = []

        # Solicitar cada expresión y su reemplazo
        for i in range(num_expresiones):
            expr_val = input(f"Enter expression {i + 1}: ")
            repl_val = input(f"Replace '{expr_val}' with: ")  # Usar comillas simples para evitar errores
            expresiones_val.append(expr_val)
            reemplazos_val.append(repl_val)

        # Buscar y reemplazar en el directorio
        contador_archivos = buscar_y_reemplazar_en_directorio(directorio_base, expresiones_val, reemplazos_val, "Output_files")

        # Mostrar el número de archivos modificados
        print("------------------------------------")
        if contador_archivos == 0:
            print("No modified files")
        else:
            # Crear un directorio de salida único si se modificó al menos un archivo
            directorio_salida = obtener_directorio_salida_unico("Output_files")
            os.makedirs(directorio_salida, exist_ok=True)
            print(f"{contador_archivos} Files modified")
        print("------------------------------------\n")

if __name__ == "__main__":
    main()

File: Multiple text file word replacer script.cpp
#include<iostream>
#include<fstream>
#include<string>
#include<vector>
#include<filesystem>
#include<algorithm>

// Reemplaza las expresiones exactas y sus versiones en mayúsculas en un archivo de texto
std::string f_reemplazar_texto(const std::string& ruta_archivo, const std::vector<std::string>& expresiones_val, const std::vector<std::string>& reemplazos_val)
{
    try {
        std::ifstream archivo_val(ruta_archivo);

        if (!archivo_val.is_open())
        {
            return "";
        }

        std::string contenido_val((std::istreambuf_iterator<char>(archivo_val)), std::istreambuf_iterator<char>());
        std::string contenido_modificado = contenido_val;

        // Bool para verificar si se realizó alguna modificación
        bool modificado_bool = false;

        // Reemplazar cada expresión y su versión en mayúsculas
        for (size_t i = 0; i < expresiones_val.size(); ++i)
        {
            const std::string& expr_val = expresiones_val[i];
            const std::string& repl_val = reemplazos_val[i];

            std::string expr_val_upper = expr_val;
            std::transform(expr_val_upper.begin(), expr_val_upper.end(), expr_val_upper.begin(), ::toupper);

            if (contenido_val.find(expr_val) != std::string::npos || contenido_val.find(expr_val_upper) != std::string::npos)
            {
                size_t pos = 0;

                while ((pos = contenido_modificado.find(expr_val, pos)) != std::string::npos)
                {
                    contenido_modificado.replace(pos, expr_val.length(), repl_val);

                    pos += repl_val.length();

                    modificado_bool = true;
                }

                pos = 0;

                while ((pos = contenido_modificado.find(expr_val_upper, pos)) != std::string::npos)
                {
                    std::string repl_val_upper = repl_val;
                    std::transform(repl_val_upper.begin(), repl_val_upper.end(), repl_val_upper.begin(), ::toupper);
                    contenido_modificado.replace(pos, expr_val_upper.length(), repl_val_upper);

                    pos += repl_val_upper.length();

                    modificado_bool = true;
                }
            }
        }

        // Devolver el contenido modificado solo si hubo cambios
        return modificado_bool ? contenido_modificado : "";
    }
    catch (const std::exception& e)
    {
        // Si ocurre un error al abrir o leer el archivo, simplemente lo ignoramos
        return "";
    }
}

// Recorre recursivamente un directorio y reemplaza las expresiones en archivos de texto
int f_buscar_reemplazar(const std::string& directorio_base, const std::vector<std::string>& expresiones_val, const std::vector<std::string>& reemplazos_val, const std::string& directorio_salida)
{
    // Contador de archivos modificados
    int contador_archivos = 0;

    for (const auto& entry : std::filesystem::recursive_directory_iterator(directorio_base))
    {
        if (entry.is_regular_file())
        {
            std::string ruta_completa = entry.path().string();

            // Intentar procesar el archivo como texto
            std::string contenido_modificado = f_reemplazar_texto(ruta_completa, expresiones_val, reemplazos_val);

            if (!contenido_modificado.empty())
            {
                // Crear la ruta de salida manteniendo la estructura de directorios
                std::string ruta_relativa = std::filesystem::relative(entry.path(), directorio_base).string();

                // Convertir las rutas a std::filesystem::path antes de usar el operador /
                std::filesystem::path subdirectorio_salida = std::filesystem::path(directorio_salida) / std::filesystem::path(ruta_relativa);

                // Crear subdirectorios si no existen
                std::filesystem::create_directories(subdirectorio_salida.parent_path());

                // Construir la ruta completa del archivo de salida
                std::filesystem::path ruta_archivo_salida = std::filesystem::path(directorio_salida) / std::filesystem::path(ruta_relativa);

                // Guardar el archivo modificado en la carpeta de salida
                std::ofstream archivo_salida(ruta_archivo_salida);

                if (archivo_salida.is_open())
                {
                    archivo_salida << contenido_modificado;

                    contador_archivos++;
                }
            }
        }
    }

    // Devolver el número de archivos modificados
    return contador_archivos;
}

// Generar un nombre de directorio único para evitar sobreescribir
std::string f_directorio_salida(const std::string& base_dir)
{
    int contador_val = 1;
    std::string directorio_salida = base_dir;

    while (std::filesystem::exists(directorio_salida))
    {
        directorio_salida = base_dir + " (" + std::to_string(contador_val) + ")";

        contador_val++;
    }

    return directorio_salida;
}

// Bucle infinito
int main()
{
    while (true)
    {
        // Solicitar el directorio de entrada
        std::string directorio_base;

        while (true)
        {
            std::cout << "Enter directory: ";
            std::getline(std::cin, directorio_base);

            if (std::filesystem::exists(directorio_base))
            {
                break;
            }

            std::cout << "Wrong directory\n";
        }

        // Solicitar el número de expresiones a modificar
        int num_expresiones_val;

        while (true)
        {
            std::cout << "Enter number of expressions to replace: ";
            std::cin >> num_expresiones_val;

            if (std::cin.good())
            {
                break;
            }

            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cout << "Wrong format\n";
        }

        // Limpiar el buffer de entrada
        std::cin.ignore();

        std::vector<std::string> expresiones_val;
        std::vector<std::string> reemplazos_val;

        // Solicitar cada expresión y su reemplazo
        for (int iter = 0; iter < num_expresiones_val; ++iter)
        {
            std::string expr_val, repl_val;
            std::cout << "Enter expression " << iter + 1 << ": ";
            std::getline(std::cin, expr_val);

            std::cout << "Replace '" << expr_val << "' with: ";
            std::getline(std::cin, repl_val);

            expresiones_val.push_back(expr_val);
            reemplazos_val.push_back(repl_val);
        }

        // Crear un directorio de salida único
        std::string directorio_salida = f_directorio_salida("Output files");
        std::filesystem::create_directory(directorio_salida);

        // Buscar y reemplazar en el directorio
        int contador_archivos = f_buscar_reemplazar(directorio_base, expresiones_val, reemplazos_val, directorio_salida);

        // Mostrar el número de archivos modificados
        std::cout << "------------------------------------\n";

        if (contador_archivos == 0)
        {
            std::cout << "No modified files\n";
        }
        else if (contador_archivos == 1)
        {
            std::cout << "1 modified file\n";
        }
        else
        {
            std::cout << contador_archivos << " modified files\n";
        }

        std::cout << "------------------------------------\n\n";
    }

    return 0;
}